saleor/graphql/product/tests/test_product.py

```diff
@@ -6373,6 +6373,144 @@ def test_product_update_variants_names(mock__update_variants_names, product_type
+            }
+        }
+    """
+    product_channel_listing = product.channel_listings.filter(
+        channel__slug=channel_USD.slug
+    ).first()
+    product_channel_listing.visible_in_listings = False
+    product_channel_listing.save()
+    product_id = graphene.Node.to_global_id("Product", product.id)
+
+    variables = {"id": product_id, "channel": channel_USD.slug}
+    response = staff_api_client.post_graphql(query, variables)
+    content = get_graphql_content(response)
+    data = content["data"]["product"]
+    assert data["id"] == product_id
+
+
+def test_unavailable_product_by_id_as_user(user_api_client, product, channel_USD):
+    query = """
+        query getProduct($id: ID!, $channel: String) {
+            product(id: $id, channel: $channel) {
+                id
+                name
+            }
+        }
+    """
+    product_channel_listing = product.channel_listings.filter(
+        channel__slug=channel_USD.slug
+    ).first()
+    product_channel_listing.visible_in_listings = False
+    product_channel_listing.save()
+    product_id = graphene.Node.to_global_id("Product", product.id)
+
+    variables = {"id": product_id, "channel": channel_USD.slug}
+    response = user_api_client.post_graphql(query, variables)
+    content = get_graphql_content(response)
+    assert content["data"]["product"] is None
+
+
+def test_unavailable_product_by_slug_as_staff(staff_api_client, product, channel_USD):
+    query = """
+        query getProduct($slug: String!, $channel: String) {
+            product(slug: $slug, channel: $channel) {
+                id
+                name
+            }
+        }
+    """
+    product_channel_listing = product.channel_listings.filter(
+        channel__slug=channel_USD.slug
+    ).first()
+    product_channel_listing.visible_in_listings = False
+    product_channel_listing.save()
+    product_id = graphene.Node.to_global_id("Product", product.id)
+
+    variables = {"slug": product.slug, "channel": channel_USD.slug}
+    response = staff_api_client.post_graphql(query, variables)
+    content = get_graphql_content(response)
+    data = content["data"]["product"]
+    assert data["id"] == product_id
+
+
+def test_unavailable_product_by_slug_as_user(user_api_client, product, channel_USD):
+    query = """
+        query getProduct($slug: String!, $channel: String) {
+            product(slug: $slug, channel: $channel) {
+                id
+                name
+            }
+        }
+    """
+    product_channel_listing = product.channel_listings.filter(
+        channel__slug=channel_USD.slug
+    ).first()
+    product_channel_listing.visible_in_listings = False
+    product_channel_listing.save()
+
+    variables = {"slug": product.slug, "channel": channel_USD.slug}
+    response = user_api_client.post_graphql(query, variables)
+    content = get_graphql_content(response)
+    assert content["data"]["product"] is None
+
+
+def test_unavailable_product_variant_by_id_as_staff(
```

Comment:
IMO, for all your tests which check the queries for staff, I would add a pytest.parametrize to cover the app case also.
Something like this should work:
```
@pytest.mark.parametrize("requestor", ["staff", "app"])
def test_exclude_products_for_shipping_method_only_products(
    requestor,
    app_api_client,
    shipping_method,
    product_list,
    staff_api_client,
    permission_manage_shipping,
):
    api = staff_api_client if requestor == "staff" else app_api_client
    ```

Reply:
Please, no `if`s in tests. Those are two separate test cases.

Reply:
So it should be tested for customer, staff_user and app.

Reply:
I added tests for app.